example-app/src/main/java/com/adyen/checkout/example/ui/googlepay/compose/SessionsGooglePayViewModel.kt

```diff
@@ -162,69 +161,34 @@ internal class SessionsGooglePayViewModel @Inject constructor(
     }
 
     override fun onFinished(result: SessionPaymentResult) {
-        updateState {
-            it.copy(uiState = SessionsGooglePayUIState.FinalResult(getFinalResultState(result)))
-        }
+        updateState { SessionsGooglePayState.FinalResult(getFinalResultState(result)) }
     }
 
     private fun getFinalResultState(result: SessionPaymentResult): ResultState = when (result.resultCode) {
         "Authorised" -> ResultState.SUCCESS
-        "Pending",
-        "Received" -> ResultState.PENDING
+        "Pending", "Received" -> ResultState.PENDING
 
         else -> ResultState.FAILURE
     }
 
     private fun onError() {
-        updateState { it.copy(uiState = SessionsGooglePayUIState.FinalResult(ResultState.FAILURE)) }
+        updateState { SessionsGooglePayState.FinalResult(ResultState.FAILURE) }
     }
 
     private fun updateState(block: (SessionsGooglePayState) -> SessionsGooglePayState) {
         _googlePayState.update(block)
     }
 
-    fun onButtonClicked() {
-        updateState {
-            it.copy(
-                uiState = SessionsGooglePayUIState.ShowComponent(componentData),
-                startGooglePay = SessionsStartGooglePayData(componentData),
-            )
-        }
+    private fun updateEvent(block: (SessionsGooglePayEvents) -> SessionsGooglePayEvents) {
+        _stateEvents.update(block)
     }
 
     fun onGooglePayLauncherResult(apiTaskResult: ApiTaskResult<PaymentData>) {
-        updateState {
-            it.copy(
-                paymentResultToHandle = SessionsGooglePayPaymentResult(
-                    componentData,
-                    apiTaskResult,
-                ),
-            )
-        }
-    }
-
-    fun onPaymentResultHandled() {
-        updateState {
-            it.copy(
-                paymentResultToHandle = null,
-            )
-        }
-    }
-
-    fun onGooglePayStarted() {
-        updateState { it.copy(startGooglePay = null) }
-    }
-
-    fun onActionConsumed() {
-        updateState { it.copy(actionToHandle = null) }
+        // TODO Call Adyen API
```

Comment:
Do you plan on completing this part or do we do it ourselves after merge?

Reply:
How is that done today?
What I've seen so far is that today, this is sent back to the compose flow, which calls the delegate's `handlePaymentResult` through the Google Pay component. I see that this creates a component state and emits it through `_componentStateFlow`. Where is this then handled?
I was hoping we could find a way to start this operation directly from the viewModel, since, after all, initiating a backend call should not need to update the UI flow.

Reply:
Yes what you said is correct. After `_componentStateFlow` emits, the component has all the data it needs to send the payment to the backend. After that you will receive one of the 3 callbacks implemented in `SessionsGooglePayViewModel`: `onAction`, `onFinished` or `onError`.

To simplify this you can remove the `onGooglePayLauncherResult` method altogether and directly call the component inside the launcher:
```kotlin
    lateinit var googlePayComponent: GooglePayComponent

    val googlePayLauncher = rememberLauncherForActivityResult(
        contract = TaskResultContracts.GetPaymentDataResult(),
        onResult = googlePayComponent::handlePaymentResult,
    )
```
WDYT?

To clarify why I didn't do this initially, I was worried about keeping an instance of the component inside the screen, to be used later. Which is why every time I needed the component I had to go back to the view model and send the component data back to fetch the component instance again. But after seeing your `lateinit` solution which we discussed earlier I think this could work.

Reply:
This works for me